### ISSUE TITLE
Cache file can get stale

### ISSUE THREAD

**Issue by ndmitchell** [haskell/ghcide#597]
I ran Ghcide over HLint. It seemingly cached compiled `.hi` objects at `C:\Users\ndmit_000\AppData\Local\ghcide`. I then upgraded `ghc-lib-parser-ex` using `cabal` and after rerunning Ghcide got warnings about HsExtendedInstances not unifying with HsExtendedInstances, talking about the old version of ghc-lib-parser-ex vs the new version. (I copied the wrong error message, so lost the actual error!) 

Should the .hi file cache be depending on the GHC package database? Or the hashes of the packages it pulls in? Having a stale disk cache runs the risk of becoming a support nightmare as people will get stuck and not know how to recover.

**Comment by ndmitchell** [comment]
CC @pepeiborra

**Comment by pepeiborra** [comment]
GHC already knows how to deal with this, so we must be doing something wrong or not performing the requisite staleness checks.

**Comment by ndmitchell** [comment]
I thought the .hi rule just did is the modification time of the .hi file newer than the .hs or not? I don't see any other safety checking. A timestamp dependency on all the package db's in scope should be sufficient (but might over-invalidate).

**Comment by pepeiborra** [comment]
ghcide delegates the safety checking to `checkOldIface`, which I believe invokes the glorious  [recompilation checking](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/compiler/recompilation-avoidance) in GHC. 

https://github.com/digital-asset/ghcide/blob/0c9a0961abbeef851b4117e6408f15a6d46eb1f1/src/Development/IDE/Core/Compile.hs#L571

**Comment by pepeiborra** [comment]
> I thought the .hi rule just did is the modification time of the .hi file newer than the .hs or not? I don't see any other safety checking. A timestamp dependency on all the package db's in scope should be sufficient (but might over-invalidate).

Specially if the package database is not sandboxed :see_no_evil:

**Comment by pepeiborra** [comment]
As per the recompilation checking notes, the interface file already tracks the package dependencies:
```
An interface file contains:


Various fingerprints:


The interface hash, which depends on the entire contents of the
interface file.  This is used to detect whether we should
update the interface on disk after recompiling the module.  If the
interface didn't change at all, then we don't want to touch the
on-disk version because that would cause make to perform more
compilations.


The ABI hash, which depends on everything that the module
exposes about its implementation: think of this as a hash of
export-list hash and decls.


The export-list hash, which depends on

The export list itself.  The export-list hash only depends on the names of the exports for the modules. The types of these exports are ignored in calculating the hash. Only a change of name or removal or addition of an export will change the hash. Not a type change of definition change.
the orphan hash, which depends on all the orphan instances/rules in the, and the orphan hashes of all orphan modules below this module in the dependency tree (see Orphans).
the package dependencies (see Package Version Changes).
```

So we must calling `checkOldIface` wrongly.

**Comment by jneira** [comment]
This maybe was fixed by #645, @ndmitchell have you hit it after that pr?

**Comment by ndmitchell** [comment]
Haven't hit it, but then upgrading a package database underneath an open session isn't a common thing to do anyway. Let's optimistically say its closed and I can always reopen if it gets hit in future.
